In [1]:
import util
import json
import gensim
from gensim.models import KeyedVectors, Word2Vec
from gensim.test.utils import get_tmpfile
import nltk
import pandas as pd
import os
import json
%reload_ext autoreload
%autoreload 2

# Creating a corpus and keyed vectors (Word2Vec) of reggaeton lyrics
In this jupyter notebook, I download the available songs 49 of the most popular reggaeton artist from the Genius website.

### Important note: most of the reggaeton lyrics are written in Spanish

1. First,  I identify a list of the most popular reggaeton artist. I found a list in 20 minutes website. I web scraped the list and obtained the name of the artists.

In [4]:
lista_20minutos = sopa = util.make_soup('https://listas.20minutos.es/lista/cantantes-de-reggaeton-y-trap-latino-426763/')

In [5]:
artists = [artist['item']['name'] for artist in json.loads(sopa.find('script', type="application/ld+json").text)['itemListElement']]

2. The next step is to use the Genius API and obtain the available songs of those artists.

In [ ]:
#songs = {artist: util.songs_dictionary(artist) for artist in artists}

In [14]:
#songs = {}

In [19]:
for artist in artists:
    if artist not in list(songs.keys()):
        songs[artist]=util.songs_dictionary(artist)

Searching for songs by Bryant Myers...

Song 1: "Vamo A Da Una Vuelta (Remix)"
Song 2: "Esclava"
Song 3: "No Eres Mía"
Song 4: "Esclava (Remix)"
Song 5: "Me Han Hablau de Ti"
Song 6: "Vamos a Vernos"
Song 7: "Bajen Pa’ Ca"
Song 8: "Momentos"
Song 9: "Tanta Falta"
Song 10: "¿Por Qué Sigues Con Él?"
Song 11: "Volvamos Hablar"
Song 12: "Viejos Tiempos"
Song 13: "Noche de Fantasía"
Song 14: "Háblame Claro"
Song 15: "Gan-Ga"
Song 16: "Ponle Música"
Song 17: "De Camino A Marte"
Song 18: "En Que País"
Song 19: "Triste (Remix)*"
Song 20: "Tú Sabes"
Song 21: "Prende Un Phillie"
Song 22: "Bryant Myers"
Song 23: "Dólares"
Song 24: "Mera Bebe"
Song 25: "Pa’ Pasar El Rato"
Song 26: "Hasta Que Me Muera"
Song 27: "Porque Sigues Con El? (Remix)"
Song 28: "En Otra Dimensión"
Song 29: "De Camino A Marte (Remix)"
Song 30: "Amor de Tres"
Song 31: "Maldad"
Song 32: "Más Que Ustedes"
Song 33: "Esclava 2 (Remix)*"
Song 34: "Tanta Falta (Remix)"
Song 35: "¿Por Que Sigues Con El? (Remix 2)"
Song 36: "Estrella*

Song 1: "HOLLYWOOD$"
Song 2: "Kolor"
Song 3: "Tinder Love"
Song 4: "Wyrwać się"
Song 5: "Ona to ziomal"
Song 6: "Na tylnych siedzeniach"
Song 7: "Lubię"
Song 8: "Moda"
Song 9: "Sam"
Song 10: "FREESTYLE [Bonus Track]"
Song 11: "Tańcz"
Song 12: "Klepsydra"
Song 13: "Nie muszę pytać"
Song 14: "Dobranoc"
Song 15: "Gomorra"
Song 16: "Cola"
Song 17: "Menago"
Song 18: "Oldschool"
Song 19: "Przez Sen"
Song 20: "6 Rano"
Song 21: "Mama, będę wielki"
Song 22: "Nie chcę"
Song 23: "Daj mi"
Song 24: "Wiele"
Song 25: "Telefony"
Song 26: "Pamiętam"
Song 27: "Kryzysowa narzeczona"
Song 28: "A4"
Song 29: "Piszę tekst"
Song 30: "Oh Love"
Song 31: "Lista Życzeń"
Song 32: "Potrzebuję"
Song 33: "Złoty Chłopak"
Song 34: "Instrukcje"
Song 35: "XXL"
Song 36: "Chłopaki z sąsiedztwa"
Song 37: "Hikikomori"
Song 38: "Stereo Typ"
Song 39: "Najlepsze lata"
Song 40: "Na dystans"
Song 41: "Immortality"
Song 42: "Na dłoni"
Song 43: "Nic nie szkodzi"
Song 44: "A co gdyby?"
Song 45: "Gdybyś tylko"
Song 46: "Chcę"
Song 47

Song 32: "Tú y Yo"
Song 33: "Fantasía*"
Song 34: "Champagne Rose*"
Song 35: "El Juego"
Song 36: "Irremplazables"
Song 37: "Nunca Imagine"
Song 38: "Una Noche"
Song 39: "Party (Remix)"
Song 40: "Mujer Prohibida"
Song 41: "Cuando Sales Sola"
Song 42: "Modales"
Song 43: "Bonnie & Clyde"
Song 44: "Chevere"
Song 45: "La Cura"
Song 46: "No Me Digas Que No"
Song 47: "¿Cómo Estás?"
Song 48: "Sensacion hermosa"
Song 49: "Soy yo"
Song 50: "Eh, Eh, Eh, Eh"
Song 51: "Hay mujeres"
Song 52: "Tu no sabes"
Song 53: "Si Ellos Supieran"
Song 54: "Macatraca"
Song 55: "Booty booty"
Song 56: "Te Confieso"
Song 57: "Suena El Dembow"
Song 58: "El Lambo, El Rari"
Song 59: "Me Calienta"
Song 60: "Solo Pienso En Ti"
Song 61: "Te Extraño"
Song 62: "Un Besito"
Song 63: "Mágico"
Song 64: "Una Vez Mas"
Song 65: "Noche y Día"
Song 66: "Como Antes*"
Song 67: "Sube Y Baja"
Song 68: "Tu Silueta"
Done. Found 68 songs.
Searching for songs by J Quiles...

Changing artist name to 'Justin Quiles'
Song 1: "No Quiero Amarte"


Song 17: "Corazón Gigante"
Song 18: "Cupido me mintio"
Song 19: "Mi Nena (Live)"
Song 20: "Mañana Sin Ti"
Song 21: "Pide Un Deseo - 826776"
Song 22: "Te Quiero Desnuda - 826109"
Song 23: "Yo Quiero Contigo - 826921"
Song 24: "Cuál Es La Diferencia"
Song 25: "Quiero Probar"
Song 26: "Cierra Los Ojos Bien"
Song 27: "Avísame"
Song 28: "Tiemblo"
Song 29: "Piensas en mi"
Song 30: "Yo Soy Asi - 826329"
Song 31: "Célula"
Song 32: "Sin Rivales"
Song 33: "Ven a Mí"
Song 34: "Yo Quiero Ver"
Song 35: "Tú Me Enloqueces"
Song 36: "El Carnaval - 2218248"
Song 37: "Qué Será"
Song 38: "Yo Soy Asi"
Song 39: "La Vida Es Cruel"
Song 40: "Yo Quiero Ver - 2043505"
Song 41: "Mi Nena"
Song 42: "Tu Cuerpo Quiero Tocar Remix"
Song 43: "Tú Te Entregas A Mí"
Song 44: "Sabes Lo Que Vas A Hacer"
Song 45: "Tu y quien mas - 826771"
Song 46: "No Me Dejes con Esa"
Song 47: "Odio (feat. nengo flow)"
Song 48: "Dime Cuando Y Donde"
Song 49: "Nunca Me Dejé"
Song 50: "Los Lobos"
Song 51: "Yo Quiero Contigo"
Song 52: "Dime 

Song 37: "Suave y Lento"
Song 38: "Prendan Los Motores"
Song 39: "Nalgona"
Song 40: "Ese Amor"
Song 41: "Un Poquito Na’ Más"
Song 42: "Sácala a Bailar"
Song 43: "Lo Que Te Gusta"
Song 44: "La Super Chapiadora"
Song 45: "Dile A El"
Song 46: "Perreame (chosen few remix)"
Song 47: "No Fue Una Noche Normal"
Song 48: "Isla del Encanto"
Song 49: "Sobredoxis de Amor"
Song 50: "Mucha Soltura"
Song 51: "Goodbye"
Song 52: "Amanece (feat. Yandel & Gadiel)"
Song 53: "¿Cómo Hago?"
Song 54: "Let’s Do It"
Song 55: "Advertios’ Están"
Song 56: "Que Te Vaya Bien (Remix)"
Song 57: "Lo Bueno Se Tarda"
Song 58: "Living in Your World"
Song 59: "Primero Báilalo"
Song 60: "Loco Contigo (Jowell)"
Song 61: "Báilalo a Lo Loco"
Song 62: "Ragga-Dub"
Song 63: "Soltura"
Song 64: "Hacerlo Así"
Song 65: "Come Back To My Crib"
Song 66: "Acomódate"
Song 67: "Un booty nuevo (feat. Yaviah)"
Song 68: "Sobredoxis"
Song 69: "Chica de Novela"
Song 70: "No Voy A Parar"
Song 71: "Tronky"
Song 72: "Posicion 4"
Song 73: "Me Prefi

Song 39: "Dañarme La Mente"
Song 40: "Tú Regresas"
Song 41: "Los Míos Ganan"
Song 42: "Estamos Clear (Solo Version)"
Song 43: "Lo Hice"
Song 44: "Alcoba (Remix)"
Song 45: "OG Goldstar (Freestyle)"
Song 46: "Nací Pa’ Ser Bueno"
Song 47: "Qué me Hicistes?"
Song 48: "Tú Regresas (Remix)"
Song 49: "En Lo Oscuro Sin Perse"
Song 50: "Barras a Lo Loco"
Song 51: "No Te Wa Mentil*"
Song 52: "Tarde o Temprano"
Song 53: "0 to 100 (Freestyle)"
Song 54: "Cheques con Dígitos"
Song 55: "Sábanas y Botellas"
Song 56: "No Apagues la Luz"
Song 57: "Alcoba"
Done. Found 57 songs.
Searching for songs by Kendo Kaponi...

Song 1: "Hola"
Song 2: "Te Amo Con Odio"
Song 3: "Amor"
Song 4: "Te Cambié"
Song 5: "Me Contagie"
Song 6: "Riendo Para No Llorar"
Song 7: "Na Na Na"
Song 8: "Para Mi Esposa"
Song 9: "Paso"
Song 10: "Dos Caras"
Song 11: "Amen"
Song 12: "Vivo Pa’ Morir"
Song 13: "No Somos Iguales"
Song 14: "No Hay Navidad Pa’ Nadie (Extended Version)"
Song 15: "Fuego"
Song 16: "Suicidio"
Song 17: "El Dios del 

Song 14: "Pégate a la Pared"
Song 15: "Cama Vacía"
Song 16: "El Primero"
Song 17: "El Rompe Discotekas"
Song 18: "Duele"
Song 19: "Los Capos Nunca Mueren"
Song 20: "Sí A La Vida"
Song 21: "Chota Cos"
Song 22: "Narco Hampon"
Song 23: "Soy El Mejor En Esto"
Song 24: "Ven Y Vuela"
Song 25: "Abre Los Ojos"
Song 26: "Amén"
Song 27: "Me puse pa ustedes"
Song 28: "Niggaz Said What"
Song 29: "Conozcan Otra Parte De Mi"
Song 30: "Las Pistolas Seguiran"
Song 31: "T.E.M.P.O."
Song 32: "Yo Corono"
Song 33: "Sín Título"
Song 34: "Nací Pa’ Ser Alguien"
Song 35: "La Mision"
Song 36: "Deja que hable el dembow"
Song 37: "Por Mí"
Song 38: "Intro"
Song 39: "Tu y quien mas"
Song 40: "Sí O Sí"
Song 41: "No Me Hables de Amor"
Song 42: "Tú Me Encantas"
Song 43: "Lo Mejor De Dos Mundos"
Song 44: "Soy El Que Tenia Que Llegar (Original)"
Song 45: "Yo Me Enamoré"
Song 46: "Freestyle*"
Song 47: "Solo Contra Todos Ustedes"
Song 48: "El Tártaro Y El Narco"
Song 49: "Cuando Tú Quieras*"
Song 50: "Natta fri"
Song 51:

This step could take a couple of hours, so a good exercise is to develop a multithreading or multiprocessing version of my functions. I plan to do so as soon as possible.

In [21]:
# I used the following lines to save the dictionary obtained. 
# Given that getting this data take a long time, you should want to do the same.

# with open('regaetton_songs.json', 'w') as fp:
#     json.dump(songs, fp)

3. Then, I create a list containing only the lyrics of the songs. Remember that the dictionary we created previously includes additional information about the songs. For now, we only need the tokens (words). Additionally, I convert all the tokens to lower. It is not necessary, but in my case, I arbitrarily decided to treat words with uppercase equal to lowercase words.

In [26]:
tokens_list = [util.clean_tokens_doc(v['lyrics']) for artist in songs.values() if artist for v in artist.values()]

In [28]:
# This is an example of how the list look like:
tokens_list[:10]

[['letra',
  '``',
  'amorfoda',
  "''",
  'estribillo',
  'quiero',
  'má',
  'nadie',
  'hable',
  'amor',
  'cansé',
  'to',
  'truco',
  'sé',
  'dolore',
  'pasé',
  'yeh',
  'yeh',
  'yeh',
  'quiero',
  'má',
  'nadie',
  'hable',
  'amor',
  'amor',
  'cansé',
  'to',
  'truco',
  'sé',
  'dolore',
  'pasé',
  'verso',
  'hoy',
  'odio',
  'secreto',
  'confieso',
  'si',
  'pudiera',
  'pidiera',
  'devuelva',
  'to',
  'beso',
  'di',
  'di',
  'palabra',
  'tiempo',
  'perdí',
  'arrepiento',
  'mil',
  'vece',
  'haber',
  'confiao',
  '¡wuh',
  'quisiera',
  'sienta',
  'siento',
  'siento',
  'quisiera',
  'ser',
  'sentimiento',
  '-miento',
  'quisiera',
  'sacarte',
  'pensamiento',
  '-miento',
  'quisiera',
  'cambiarle',
  'final',
  'cuento',
  'toa',
  'barra',
  'trago',
  'sido',
  "testigo'",
  'dolor',
  'causaste',
  'to',
  'hiciste',
  'conmigo',
  'infeliz',
  'amor',
  'aún',
  'supera',
  'ahora',
  'camina',
  'solo',
  'nadie',
  'toa',
  "acera'",
  '

In [54]:
len(tokens_list)

3126

4. The next step is to train the Word2Vec model. I used the Gensim library for simplicity. In my case, I chose 300 as the dimension of the vectors and five as a minimum number of appearances of words. These are the same parameters selected by Rachel Tatman to train a Word2Vec model based on the Spanish Billion Words Corpus (https://www.kaggle.com/rtatman/pretrained-word-vectors-for-spanish). I used those pre-trained vectors for comparison in the "similaritiesw2v" notebook. 
    
You can check the Gensim documentation here: https://radimrehurek.com/gensim/models/word2vec.html#gensim.models.word2vec.Word2Vec

And if you want to know more about this kind of embeddings you can see this famous paper: https://arxiv.org/abs/1301.3781

In [69]:
model = Word2Vec(tokens_list, size=300, min_count=5)

In [71]:
words = list(model.wv.vocab)
len(words)

13950

5. Once we finished the training part, I store the word vectors to use them in the future. This model did not take a long time to train because the reggaeton corpus is tiny compared to corpora that are usually used to train Word2Vec models. You can see that the corpus produced a vocabulary of only 13,950 words, while the model used for comparison produced 1,000,653 words — this a fundamental limitation to take into account.

In [33]:
word_vectors = model.wv

In [72]:
fname = get_tmpfile(os.getcwd()+"\REGGAETON-vectors-300-min5.kv")
word_vectors.save(fname)
word_vectors = KeyedVectors.load(fname, mmap='r')

##### I will use this keyed vectors in the "similaritiesw2v" jupyter notebook to analyze the most similar words to specific terms.